### ISSUE TITLE
Contravariance + interfaces for Whens?

### ISSUE THREAD

**Issue by ashic** [yreynhout/Projac#12]
Does projac currently support interfaces and / or contravariance in interfaces? Example:

```
public interface Envelope { string StreamId {get;} }
public interface Envelope<out T> : Envelope { T Message {get;} }

public class EnvelopeImpl<T> : Envelope<T>{
   ...
}


return new SqlProjectionBuilder()
                .When<Envelope<SomeEvent>>(x =>
                {
                    Console.WriteLine("there");
                    return new SqlNonQueryCommand(...);
                })
```

It seems like When&lt;Envelope&gt; and When&lt;Envelope&lt;SomeEvent&gt;&gt; are not triggered. Only When&lt;EnvelopeImpl&lt;SomeEvent&gt;&gt; is. That wouldn't be a problem if When&lt;EnvelopeImple&lt;SomeEventInterface&gt;&gt; worked, but that doesn't either.

The scenario I'm targeting is this:

```
public class SomeEvent1 : SomethingStarted
{
}

public class SomeEvent2 : SomethingStarted
{
}

public interface SomethingStarted
{
    //some props
}
```

And I'm looking to have a projection for all Envelope&lt;SomethingStarted&gt; events. Is this possible?

**Comment by yreynhout** [comment]
Sorry for the late response. This will require a custom SqlProjector implementation - which I assume you are using to do the actual dispatching of "Whens" (aka SqlProjectionHanders). I doubt it can be generalized because everybody brings their own conventions (e.g. what if I brought a Envelope<TMessage, TSomethingElse>?) and assumes their dispatch order will be respected. In such cases it's just better to code out the dispatching yourself. I will post a gist of a SqlProjector that does what you want but just gimme some time.

**Comment by ashic** [comment]
I can easily see how that can be handled with a custom projector. However, it might be possible to do handler resolution that honours interface contravariance in a generic manner. Currently GetHandlers() simply looks up the handlers in a dictionary. They key for the dictionary may be a wrapper around Type with equality implementations to cater for contravariance.  I'll play around a bit.

**Comment by ashic** [comment]
Right...so put together this little thing: https://github.com/ashic/contravariant-handling 

We should be able to make that injectable. Performance wise, looping through handlers v dictionary lookup won't have much impact, but making it injectable makes sense. If it looks ok, then I'll adapt it for projac and send in a PR. Sound good?

**Comment by ashic** [comment]
Just put out a nuget package (contra). It's effectively two classes similar to the above code. But this one's generic. I've got a few places where I'll use this, hence the package. Thoughts?

https://github.com/heartysoft/contra

**Comment by yreynhout** [comment]
In general, I agree, Projac needs to support these scenarios, preferably OOTB. I'd like to postpone any implementation of it, until I've dealt with deprecating all the client lib specific packages and projects (#14). Makes testing it easier on my end. I'm not going to take a dependency on contra though. I'll just be copying the bits that suite me. I want to play around in this area myself mainly because I've got some learning to do.

If there's no overhead associated with this, I could choose to support this as the default, or make it, as you suggest, an injectable strategy. Still, in the back of my mind, I think the matching is a lazy, one time operation, on a per input message type basis, for which the answer remains stable given a fixed set of handlers. If this would be the only strategy, because, honestly, I can't think of any other, than it might make more sense to move it to the XXXProjectionHandler (a.k.a. your HandlerEntry). Need to ponder on this ...

The only downside is that I need to resolve this at runtime, and if I want to cache the result of the resolution, I can't make any assumptions about whether people are using the projection from one or multiple threads, which kinda sucks ATM. Unless, I drop the caching part, ... would have to measure cost vs benefit - sounds very marginal at this point.

**Comment by yreynhout** [comment]
@ashic work has started on this over here: https://github.com/yreynhout/Projac/tree/feature-contrapatch/src/Projac - it seems, from my limited testing, that the dispatching you are looking for can be achieved with the use of IsAssignableFrom (or IsInstanceOfType in my case) instead of the manual matching. I still have to port all your test scenarios to make sure + beef up the testing on my end, though. Will keep you informed about progress.

**Comment by ashic** [comment]
check out the contra tests. the demo had a few bugs, and contra has some additional test cases.

Glad to see work's started :)

**Comment by yreynhout** [comment]
@ashic I've been able to port all the TypeRegistry tests except for one. That is, the last one (which you probably knew in advance). You had me baffled there for a minute. But then I realized it's illegal C# to begin with, since there's no way to go from a ConcreteEnvelope<PersonRegisteredEvent> to a ConcreteEnvelope<RegisteredEvent> (try writing a method that does). Sure, your dispatching works, but that's just because your matching logic will present a ConcreteEnvelope<RegisteredEvent> handler as a match. Yet, looking closer at the lambda, x is of type System.Object, which, in handlers for this purpose, is totally useless. The entire point is to have x be of type ConcreteEnvelope<RegisteredEvent>. Something which a type of ConcreteEnvelope<PersonRegisteredEvent> will never be. No point in supporting that, unless you'd manually convert one to the other. So, unless, you can come up with a test that shows me automagic conversion, I'm not going to support this particular case.

https://github.com/yreynhout/Projac/blob/feature-contrapatch/src/Projac.Tests/ResolveTests.cs

**Comment by ashic** [comment]
I don't get it. RegisteredEvent is an abstraction. PersonRegistered is an implementation. If I have a listener on ConcreteEnvelope<RegisteredEvent>, if a ConcreteEnvelope<PersonRegistered> fires, I should get it, no? Yes, in the lambda, x=> ... , x will be of type ConcreteEnvelope<Registered>, but it should be possible.

**Comment by ashic** [comment]
If I remember correctly, the envelope interface being contravariant automatically takes care of that.

**Comment by yreynhout** [comment]
The envelope interface, yes. The concrete envelope, nope.

**Comment by yreynhout** [comment]
ConcreteEnvelope&lt;PersonRegisteredEvent&gt; can not be converted to ConcreteEnvelope&lt;RegisteredEvent&gt;. It's not possible, try it.

**Comment by ashic** [comment]
I'm not saying it should be convertable. I'm talking about what makes sense from a subscriber's point of view. If I want to subscribe to the envelope of an abstraction, and envelope of an implementation of that abstraction should reach me, no?

**Comment by yreynhout** [comment]
In C#, unless there's an automatic conversion, no, what you're asking for is not possible, even if it sounds reasonable. The way to work around that limitation is to make the envelope an abstraction as well (where the contravariance exists). That's the best we can do at the moment.

**Comment by yreynhout** [comment]
I should also mention I'm not a big fan of messages with multiple interfaces. This whole use case is a bit far fetched from my point of view. But then we're venturing into the land of preference ...

**Comment by yreynhout** [comment]
The bad news we already knew: http://stackoverflow.com/questions/2541467/why-does-c-sharp-4-0-not-allow-co-and-contravariance-in-generic-class-types

The good news is that once I land custom handler resolvers, you'll be able to manually do the conversion if you feel like it.

**Comment by ashic** [comment]
I think the test in contra would easily pass given that it's the contravariance of the envelope interface that makes it pass. I'd expect that to pass for your tests as well, as long as there's a contravariant envelope interface.

**Comment by yreynhout** [comment]
It passes if you expect a Envelope&lt;RegisteredEvent&gt; handler to receive an instance of ConcreteEnvelope&lt;PersonRegisteredEvent&gt;. Since contravariance does not exist at class level, you'll never be able to automatically receive an instance of ConcreteEnvelope&lt;PersonRegisteredEvent&gt; in a ConcreteEnvelope&lt;RegisteredEvent&gt; handler. If you think it's possible, show me a test that proves that, because I've just written one that proves you get an InvalidCastException.
```
    [TestFixture]
    public class ResolveTests
    {
        [Test]
        public void ItCanNotBeDone()
        {   //where Message1 : IMessage
            var handler = HandlerForType<MessageEnvelope<IMessage>>(_ => new SqlNonQueryCommand[0]);
            Assert.DoesNotThrow(() => 
                { var _ = handler.Handler(new MessageEnvelope<Message1>()).ToArray(); });
        }

        private static SqlProjectionHandler HandlerForType<TMessage>(Func<TMessage, SqlNonQueryCommand[]> handler)
        {
            return new SqlProjectionHandler
                (
                typeof (TMessage),
                message => handler((TMessage) message)
                );
        }
    }
```
The problem isn't so much the resolution of the handler. We can always decide that a ConcreteEnvelope&lt;RegisteredEvent&gt; handler matches when we observe an ConcreteEnvelope&lt;PersonRegisteredEvent&gt; message because that's entirely up to our own conventions on what matches and what doesn't. The problem is that we can't cast an instance of ConcreteEnvelope&lt;PersonRegisteredEvent&gt; to ConcreteEnvelope&lt;RegisteredEvent&gt;. The handler is never able to observe the ConcreteEnvelope&lt;PersonRegisteredEvent&gt; message instance as a ConcreteEnvelope&lt;RegisteredEvent&gt; message. The only way to be able to do that is if we brought our own conversion logic to convert one into the other (but then it'd be a different message instance).

**Comment by ashic** [comment]
This is what I'm referring to:

https://gist.github.com/ashic/86757cf63b775f659d05 

Basic interface contravariance would take care of that.

**Comment by yreynhout** [comment]
Again - we're going around in circles here - that only works because your handler, Action<object>, doesn't give you a typed version of the message (parameter) passed in of the type you registered the handler for, which IS the entire point of a handler in Projac. I don't want you to be doing casting yourself in your handlers, what would be the point. If you want to cast to Envelope&lt;Registered&gt; you could just as well subscribe to (aka register) Envelope&lt;Registered&gt; in the first place. So, no, basic interface contravariance is not taking care of it (for the Projac needs).

If you want to be doing the casting in your handler yourself and/or non-automatic conversion of the message, then I don't think Projac's SqlProjectionBuilder or the Async-/SqlProjector will be of much use. The only thing left at that point is a SqlProjection and SqlProjectionHandler and building your own registrar and dispatcher.

**Comment by yreynhout** [comment]
Fixed up to the point where I'm comfortable with the solution (even if I don't support that last scenario).
